test/unit/workDiv/src/WorkDivForKernelTest.cpp

```diff
@@ -0,0 +1,253 @@
+        CHECK(workDiv == WorkDiv{Vec{threadsPerGridTestValue}, Vec{1}, Vec{1}});
+        // Test a new 1D workdiv. Threads per block can not be larger than 1 for CPU. Hence 2 is not valid.
+        auto const& workDiv1DUsingInitList = WorkDiv{Vec{threadsPerGridTestValue / 2}, Vec{2}, Vec{1}};
+        auto const isWorkDivValidForCPU
+            = alpaka::isValidWorkDivKernel<Acc>(dev, bundeledKernel, workDiv1DUsingInitList);
+        CHECK(isWorkDivValidForCPU == false);
+    }
+    else
+    {
+        throw std::invalid_argument("Acc type is not among tested Accs.");
+    }
+}
+
+using TestAccs2D = alpaka::test::EnabledAccs<alpaka::DimInt<2u>, std::uint32_t>;
+
+TEMPLATE_LIST_TEST_CASE("getValidWorkDivForKernel.2D", "[workDivKernel]", TestAccs2D)
+{
+    using Acc = TestType;
+    using Idx = alpaka::Idx<Acc>;
+    using Dim = alpaka::Dim<Acc>;
+    using Vec = alpaka::Vec<Dim, Idx>;
+    using WorkDiv = alpaka::WorkDivMembers<Dim, Idx>;
+    auto const platform = alpaka::Platform<Acc>{};
+    auto const dev = alpaka::getDevByIdx(platform, 0);
+
+    TestKernelWithManyRegisters kernel;
+    // A random value
+    size_t val(200ul);
+    auto const& bundeledKernel = alpaka::KernelBundle(kernel, val);
+
+    // Get hard limits for test
+    auto const& props = alpaka::getAccDevProps<Acc>(dev);
+    Idx const threadsPerGridTestValue = props.m_blockThreadCountMax * props.m_gridBlockCountMax;
+
+    auto const threadElemExtent = Vec::all(1);
+    auto gridThreadExtent = Vec::all(1);
+    if(threadsPerGridTestValue >= 2)
+        gridThreadExtent = Vec{2u, threadsPerGridTestValue / 2u};
+
+    // Test getValidWorkDivForKernel function for threadsPerGridTestValue threads per grid.
+    auto const workDiv
+        = alpaka::getValidWorkDivForKernel<Acc>(dev, bundeledKernel, gridThreadExtent, threadElemExtent);
+
+    // Test isValidWorkDivKernel function
+    auto const isValid = alpaka::isValidWorkDivKernel<Acc>(dev, bundeledKernel, workDiv);
+    CHECK(isValid == true);
+
+    if constexpr(alpaka::accMatchesTags<Acc, alpaka::TagGpuCudaRt>)
+    {
+        // Expected valid workdiv for this kernel. These values might change depending on the GPU type and compiler
+        // therefore commented out. Number of registers used by kernel might limit threads per block value differently
+        // for different GPUs. CHECK(workDiv == WorkDiv{Vec{8, 729444}, Vec{1, 736}, Vec{1, 1}});
+
+        // Get calculated threads per block from the workDiv that was found by examining kernel function
+        auto const threadsPerBlock = workDiv.m_blockThreadExtent.prod();
+        // Get hard limits
+        auto const threadsPerBlockLimit = props.m_blockThreadCountMax;
+
+        // Depending on the GPU type or the compiler the test below might fail because threadsPerBlock can be equal to
+        // threadsPerBlockLimit, which is the max device limit.
+        CHECK(threadsPerBlock < static_cast<Idx>(threadsPerBlockLimit));
```

Comment:
```suggestion
        CHECK(threadsPerBlock <= static_cast<Idx>(threadsPerBlockLimit));
```

Using the maximum value should be fine.

Reply:
I wanted to show that function attributes for this specific kernel function is less than the device properties. This is risky in a test but this is the idea of fixing the problem.